File: clients/python/src/cien_agent_sdk/admin/sync_source_definitions.py
from __future__ import annotations

from typing import Any

from ..base import EndpointGroup
from ..utils import drop_none


class AdminSyncSourceDefinitionsAPI(EndpointGroup):
    """/api/admin/sync-source-definitions endpoints."""

    def list(self, *, is_active: bool | None = None) -> list[dict[str, Any]]:
        return self._get("/api/admin/sync-source-definitions", params=drop_none({"is_active": is_active}))

    def get(self, definition_id: int) -> dict[str, Any]:
        return self._get(f"/api/admin/sync-source-definitions/{definition_id}")

    def get_by_source_type(self, source_type: str) -> dict[str, Any]:
        return self._get(f"/api/admin/sync-source-definitions/source-type/{source_type}")

    def create(
        self,
        *,
        display_name: str,
        source_type: str,
        meltano_plugin_name: str,
        env_prefix: str,
        required_settings: list[Any] | None = None,
        is_active: bool = True,
    ) -> dict[str, Any]:
        return self._post(
            "/api/admin/sync-source-definitions",
            json={
                "display_name": display_name,
                "source_type": source_type,
                "meltano_plugin_name": meltano_plugin_name,
                "env_prefix": env_prefix,
                "required_settings": required_settings or [],
                "is_active": is_active,
            },
        )

    def update(self, definition_id: int, payload: dict[str, Any]) -> dict[str, Any]:
        return self._patch(f"/api/admin/sync-source-definitions/{definition_id}", json=payload)

    def delete(self, definition_id: int) -> None:
        self._delete(f"/api/admin/sync-source-definitions/{definition_id}")

File: clients/node/src/admin/sync-source-definitions.ts
import { EndpointGroup } from '../base'
import { dropNullish } from '../utils'

export class AdminSyncSourceDefinitionsAPI extends EndpointGroup {
  list(params: { is_active?: boolean } = {}) {
    return super.requestGet<Array<Record<string, unknown>>>('/api/admin/sync-source-definitions', {
      params: dropNullish(params),
    })
  }

  get(definitionId: number) {
    return super.requestGet<Record<string, unknown>>(`/api/admin/sync-source-definitions/${definitionId}`)
  }

  getBySourceType(sourceType: string) {
    return super.requestGet<Record<string, unknown>>(
      `/api/admin/sync-source-definitions/source-type/${sourceType}`
    )
  }

  create(payload: {
    display_name: string
    source_type: string
    meltano_plugin_name: string
    env_prefix: string
    required_settings?: unknown[]
    is_active?: boolean
  }) {
    return this.requestPost<Record<string, unknown>>('/api/admin/sync-source-definitions', {
      json: {
        ...payload,
        required_settings: payload.required_settings ?? [],
        is_active: payload.is_active ?? true,
      },
    })
  }

  update(definitionId: number, payload: Record<string, unknown>) {
    return this.requestPatch<Record<string, unknown>>(
      `/api/admin/sync-source-definitions/${definitionId}`,
      { json: payload }
    )
  }

  async delete(definitionId: number) {
    await super.requestDelete(`/api/admin/sync-source-definitions/${definitionId}`)
  }
}
